internal/internal.go

```diff
@@ -612,39 +620,123 @@ func getParsedPathAndQuery(req *ext_authz.CheckRequest) ([]interface{}, map[stri
 	return parsedPathInterface, parsedQueryInterface, nil
 }
 
-func getParsedBody(req *ext_authz.CheckRequest) (interface{}, bool, error) {
-	body := req.GetAttributes().GetRequest().GetHttp().GetBody()
+func getParsedBody(req *ext_authz.CheckRequest, parsedPath []interface{}, protoDescriptor string) (interface{}, bool, error) {
 	headers := req.GetAttributes().GetRequest().GetHttp().GetHeaders()
 
-	if body == "" {
-		return nil, false, nil
-	}
-
-	if val, ok := headers["content-length"]; ok {
-		cl, err := strconv.ParseInt(val, 10, 64)
-		if err != nil {
-			return nil, false, err
-		}
-
-		if cl != -1 && cl > int64(len(body)) {
-			return nil, true, nil
-		}
-	}
-
 	var data interface{}
 
 	if val, ok := headers["content-type"]; ok {
 		if strings.Contains(val, "application/json") {
+
+			body := req.GetAttributes().GetRequest().GetHttp().GetBody()
+
+			if body == "" {
+				return nil, false, nil
+			}
+
+			if val, ok := headers["content-length"]; ok {
+				cl, err := strconv.ParseInt(val, 10, 64)
+				if err != nil {
+					return nil, false, err
+				}
+				if cl != -1 && cl > int64(len(body)) {
+					return nil, true, nil
+				}
+			}
+
 			err := util.Unmarshal([]byte(body), &data)
 			if err != nil {
 				return nil, false, err
 			}
+		} else if strings.Contains(val, "application/grpc") {
+
+			if protoDescriptor == "" {
+				return nil, false, nil
+			}
+
+			rawbody := req.GetAttributes().GetRequest().GetHttp().GetRawBody()
+
+			if len(rawbody) <= 0 {
+				return nil, false, fmt.Errorf("invalid raw body")
+			}
+			if len(parsedPath) <= 1 {
```

Comment:
Why does the parsed path need to have a length of 2 ?

Reply:
Because we need to know the input type to decode. To find the input type to decode we need have the package + method, 
that information its present in parsedpath[1]